src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -2023,7 +2024,14 @@ void MergeTreeDataMergerMutator::writeWithProjections(
     std::vector<SquashingTransform> projection_squashes;
     for (size_t i = 0, size = projections_to_build.size(); i < size; ++i)
     {
-        projection_squashes.emplace_back(65536, 65536 * 256);
+        // If the parent part is an in-memory part, squash projection output into one block and
+        // build in-memory projection because we don't support merging into a new in-memory part.
+        // Otherwise we split the materialization into multiple stages similar to the process of
+        // INSERT SELECT query.
+        if (new_data_part->getType() == MergeTreeDataPartType::IN_MEMORY)
+            projection_squashes.emplace_back(std::numeric_limits<size_t>::max(), std::numeric_limits<size_t>::max());
+        else
+            projection_squashes.emplace_back(65536, 65536 * 256);
```

Comment:
Now, when I look at this row, wouldn't it better use `min_insert_block_size_rows` and `min_insert_block_size_bytes` from context.settings here?